ntp-proto/src/filter.rs

```diff
@@ -0,0 +1,315 @@
+
+        for tuple in self.register.iter_mut() {
+            tuple.dispersion += dispersion_correction;
+
+            std::mem::swap(&mut current, tuple);
+        }
+    }
+
+    fn sort_by_delay(&mut self) {
+        self.register.sort_by(|t1, t2| {
+            t1.delay
+                .partial_cmp(&t2.delay)
+                .unwrap_or_else(|| panic!("got a NaN"))
+        });
+    }
+
+    fn valid_tuples(&self) -> &[FilterTuple] {
+        let num_invalid_tuples = self
+            .register
+            .iter()
+            .rev()
+            .take_while(|t| t.is_dummy())
+            .count();
+
+        let num_valid_tuples = self.register.len() - num_invalid_tuples;
+
+        &self.register[..num_valid_tuples]
+    }
+
+    /// #[no_run]
+    ///                     i=n-1
+    ///                     ---     epsilon_i
+    ///      epsilon =       \     ----------
+    ///                      /        (i+1)
+    ///                     ---     2
+    ///                     i=0
+    /// Invariant: the register is sorted wrt delay
+    fn dispersion(self) -> NtpDuration {
+        self.register
+            .iter()
+            .enumerate()
+            .map(|(i, t)| t.dispersion / 2i64.pow(i as u32 + 1))
+            .fold(NtpDuration::default(), |a, b| a + b)
+    }
+
+    /// #[no_run]
+    ///                          +-----                 -----+^1/2
+    ///                          |  n-1                      |
+    ///                          |  ---                      |
+    ///                  1       |  \                     2  |
+    ///      psi   =  -------- * |  /    (theta_0-theta_j)   |
+    ///                (n-1)     |  ---                      |
+    ///                          |  j=1                      |
+    ///                          +-----                 -----+
+    ///
+    /// Invariant: the register is sorted wrt delay
+    fn jitter(self, s: &System) -> f64 {
+        let register = self.valid_tuples();
+
+        // for jitter, only the valid tuples are considered
+        let smallest_delay = match register.get(0) {
+            None => unreachable!(
+                "there must be at least one valid tuple, this is guaranteed by clock_filter"
+            ),
+            Some(t) => t,
+        };
+
+        let root_mean_square = register
+            .iter()
+            .map(|t| (t.offset - smallest_delay.offset).to_seconds().powi(2))
+            .sum::<f64>()
+            .sqrt();
+
+        // root mean square average (RMS average). - 1 to exclude the smallest_delay
+        let jitter = root_mean_square / (register.len() - 1) as f64;
+
+        // In order to ensure consistency and avoid divide exceptions in other
+        // computations, the psi is bounded from below by the system precision
+        // s.rho expressed in seconds.
+        jitter.max(s.precision)
+    }
+}
+
+pub struct System {
+    precision: f64,
+    poll: NtpDuration,
+    leap_indicator: NtpLeapIndicator,
+}
```

Comment:
these structs are not complete. Figuring out what the correct type should be is a matter of trial and error, so I'm adding fields as needed.